compiler/rustc_passes/src/check_attr.rs

```diff
@@ -433,23 +433,21 @@ impl<'tcx> CheckAttrVisitor<'tcx> {
 
     /// Checks that `#[optimize(..)]` is applied to a function/closure/method,
     /// or to an impl block or module.
-    // FIXME(#128488): this should probably be elevated to an error?
-    fn check_optimize(&self, hir_id: HirId, attr: &Attribute, target: Target) {
-        match target {
+    fn check_optimize(&self, hir_id: HirId, attr: &Attribute, span: Span, target: Target) {
+        let is_valid = matches!(
+            target,
             Target::Fn
-            | Target::Closure
-            | Target::Method(MethodKind::Trait { body: true } | MethodKind::Inherent)
-            | Target::Impl
-            | Target::Mod => {}
-
-            _ => {
-                self.tcx.emit_node_span_lint(
-                    UNUSED_ATTRIBUTES,
-                    hir_id,
-                    attr.span,
-                    errors::OptimizeNotFnOrClosure,
-                );
-            }
+                | Target::Closure
+                | Target::Method(MethodKind::Trait { body: true } | MethodKind::Inherent)
+                | Target::Impl
+                | Target::Mod
+        );
+        if !is_valid {
```

Comment:
Problem: cf. https://github.com/rust-lang/rust/pull/128943#issuecomment-2417543309 and https://hackmd.io/-xvHdfXOQ5ucym2srg9eFA?view#inlinenever-does-not-work-for-async-functions-rust129347

> I'll just jot a follow-up note here that one place where we probably do want to continue accepting `#[optimize]` is on async blocks, with the semantics that the attribute is applied to the poll method of the generated future.

This should have a test.